util/mSCGroup.py

```diff
@@ -0,0 +1,305 @@
+
+        if len(anndata.obsp.keys()) > 0:
+            obsp_uri = self.write_annotation_matrices(anndata.obsp, "obsp")
+            base_group.add(uri=obsp_uri, relative=False, name="obsp")
+
+        if len(anndata.varp.keys()) > 0:
+            varp_uri = self.write_annotation_matrices(anndata.varp, "varp")
+            base_group.add(uri=varp_uri, relative=False, name="varp")
+
+        # ----------------------------------------------------------------
+        if self.verbose:
+            print("  FINISH WRITING %s" % (self.uri))
+
+        base_group.close()
+
+    # ----------------------------------------------------------------
+    def write_X(self, anndata):
+        X_uri = os.path.join(self.uri, "X")
+        tiledb.group_create(X_uri)
+        X_group = tiledb.Group(X_uri, "w")
+
+        X_data_uri = os.path.join(X_uri, "data")
+        if self.verbose:
+            print("    START  WRITING %s" % (X_data_uri))
+
+        # Here we do not use tiledb.from_numpy, so that we can have more control over the schema.
+
+        obs_dim, var_dim = np.meshgrid(anndata.obs.index, anndata.var.index)
+
+        dom = tiledb.Domain(
+            tiledb.Dim(name="obs_id", domain=(None, None), dtype="ascii"),
+            tiledb.Dim(name="var_id", domain=(None, None), dtype="ascii"),
+        )
+        att = tiledb.Attr("data")
+        sch = tiledb.ArraySchema(domain=dom, attrs=(att,), sparse=True)
+        tiledb.Array.create(X_data_uri, sch)
+
+        with tiledb.open(X_data_uri, "w") as A:
+            A[np.ravel(obs_dim), np.ravel(var_dim)] = anndata.X.flatten()
+
+        X_group.add(uri=X_data_uri, relative=False, name="data")
+        if self.verbose:
+            print("    FINISH WRITING %s" % (X_data_uri))
+
+        X_group.close()
+
+        return X_uri
+
+    # ----------------------------------------------------------------
+    def write_obs_or_var(self, obs_or_var_data, obs_or_var_name):
+        """
+        First argument is anndata.obs or anndata.var; second is "obs" or "var".  In the reference
+        pbmc3k_processed dataset, these are of type pandas.core.frame.DataFrame. In further
+        testing we may need to switch on the datatype.
+        """
+    
+        offsets_filters = tiledb.FilterList(
+            [tiledb.PositiveDeltaFilter(), tiledb.ZstdFilter(level=-1)]
+        )
+        coords_filters = tiledb.FilterList([tiledb.ZstdFilter(level=-1)])
+        dim_filters = tiledb.FilterList([tiledb.ZstdFilter(level=-1)])
+        attr_filters = tiledb.FilterList([tiledb.ZstdFilter(level=-1)])
+
+        obs_or_var_uri = os.path.join(self.uri, obs_or_var_name)
+        if self.verbose:
+            print("    START  WRITING %s" % (obs_or_var_uri))
+
+        tiledb.from_pandas(
+            uri=obs_or_var_uri,
+            dataframe=obs_or_var_data,
+            name=obs_or_var_name,
+            sparse=True,
+            allows_duplicates=False,
+            coords_filters=coords_filters,
```

Comment:
hmm I believe setting `coords_filters` will be deprecated in core, and there is already a `DeprecationWarning` in TileDB-Py for this, but if this is to keep the schema/array consistent with what we already have then ignore this comment.

Reply:
@nguyenv thanks -- this was a copy from another file I encountered while assembling this; perhaps @aaronwolen has insight -- I'll test this both with & without `coords_filters` & if all is well, I'll remove this ...

Reply:
@nguyenv there was `os.path.join` in another file, and this `+` here ... i went with `os.path.join` for consistency along with a portability note in `apis/python/README.md`

Reply:
https://github.com/single-cell-data/TileDB-SingleCell/pull/5/commits/1acf4e657433fee86f38bdd8fd887aeaba185f0f

Reply:
That's news to me but if that's the case I'll remove the option as well.